### ISSUE TITLE
Configuration update prevents pool rollback

### ISSUE THREAD

**Issue by behlendorf** [zfsonlinux/zfs#2452]
While debugging a damaged pool I came across a case where `zpool import -F` fails even though the txg is intact.  The interesting thing here is that everything appears to be working as designed and this simply may not be a failure mode which was fully considered.  Let me try and explain.

When a configuration change is made to a pool new labels are written out to several of the vdevs.  The new label contains the updated configuration and the txg number in which the label was written.  This couples the configuration to a specific version of the pool which is a good thing.

Now when a pool is imported the function `vdev_uberblock_load` will attempt to locate the best uberblock from all the possible vdevs.  The best uberblock is roughly defined as the one with the largest txg although there are a few caviots.  Once an uberblock and vdev are selected the label is read and certain sanity checks are performed.

One of those checks is to verify that the label txg is less than the uberblock txg and therefore contains a correct configuration.  There's a nice comment describing this case in the code.

```
        /*
         * It's possible that the best uberblock was discovered on a label
         * that has a configuration which was written in a future txg.
         * Search all labels on this vdev to find the configuration that
         * matches the txg for our uberblock.
         */
```

The problem is that this has the side effect of disabling pool rollback to a txg when this vdev is selected.  However, this doesn't have to be the case.  It's only true because the `vdev_uberblock_load` function does not try all the other vdevs with equivalent uberblocks for that txg.  Those vdevs may have older labels and allowing `vdev_uberblock_load` to check them would make the rollback more robust.

Additionally, I think a reasonable case can be made that when either the `-F` of `-X` rollback options are given we should be less strict about this constraint.  Using a label which is slightly out of date isn't always strictly going to be a problem.  And in fact this exact case becomes more and more likely the farther you rollback the pool.  Attempting to use this newer label when in recovery mode is going to far preferable to the alternative of pool loss.  For example, the pool mentioned above was able to be recovered by simply disabling this check.

**Comment by richardk3nt** [comment]
I'm trying to recover some deleted files using the rollback to a specific transaction group.  I snagged the ZFS code via git, made the change mentioned above and compiled but still get the "cannot import 'tank': one or more devices is currently unavailable" error.  Are there additional code changes necessary now to have this work?  I've tried this on Ubuntu 12.04.5 (precise) if that matters any.

**Comment by kpande** [comment]
@richardk3nt it's possible that your system doesn't need those changes, however, this isn't a support forum. Please try list.zfsonlinux.org -> zfs-discuss and the users on there will be able to help you.

**Comment by mailinglists35** [comment]
@kpande ok, not a support forum, but just giving feedback that the patch does not have the intended effect.
@behlendorf, I also am trying to import an exported healthy pool (no configuration change, just deleted some stuff, exported pool, wanting to recover) but getting the same error message.

**Comment by behlendorf** [comment]
@mailinglists35 even with the patch there's no guarantee you'll be able to rollback more than a few TXGs.  At that point ZFS can start overwriting previously freed blocks.

**Comment by mailinglists35** [comment]
@behlendorf 
with the below testcase, should the file recoverable? because on my case, it gives the same error message even if I try to use the previous txg.

    root@linux:~# truncate -s 2GiB /tmp/deleteme
    root@linux:~# zpool create test /tmp/deleteme
    root@linux:~# zfs mount
    test                            /test
    root@linux:~# cp /bin/true /test/
    root@linux:~# zpool export test
    root@linux:~# zpool import -d /tmp test
    root@linux:~# rm /test/true
    /bin/rm: remove regular file ‘/test/true’? y
    root@linux:~# zpool export test
    root@homerouter:~# zdb -ul /tmp/deleteme|grep txg
    [...]
        txg: 27
            create_txg: 4
            txg = 4
            txg = 5
            txg = 11
            txg = 13
            txg = 16
            txg = 18
            txg = 20
            txg = 22
            txg = 23
            txg = 24
            txg = 27
    root@linux:~# zpool import -o readonly=on -T 24 -d /tmp test
    cannot import 'test': one or more devices is currently unavailable

**Comment by behlendorf** [comment]
@mailinglists35 yes, I'd expect that to work but you may need to also provide -F.

**Comment by mailinglists35** [comment]
the manual page says that -F and -X are implied when doing -T. Adding -F does not produce any change.

**Comment by mailinglists35** [comment]
and I'm pretty sure I have modified the vdev_label.c file:

root@linux:~# grep -C4 'label_txg > best_txg'  /var/lib/dkms/zfs/0.6.5.7/build/module/zfs/vdev_label.c
                            ZPOOL_CONFIG_POOL_TXG, &label_txg);
                        if ((error || label_txg == 0) && !config) {
                                config = label;
                                break;
                        } else if (label_txg > best_txg) {
                                best_txg = label_txg;
                                nvlist_free(config);
                                config = fnvlist_dup(label);
                        }


root@linux:~# grep -C4 'label_txg > best_txg'  /var/lib/dkms/zfs/0.6.5.7/source/module/zfs/vdev_label.c
                            ZPOOL_CONFIG_POOL_TXG, &label_txg);
                        if ((error || label_txg == 0) && !config) {
                                config = label;
                                break;
                        } else if (label_txg > best_txg) {
                                best_txg = label_txg;
                                nvlist_free(config);
                                config = fnvlist_dup(label);
                        }
`

**Comment by mailinglists35** [comment]
ah, it seems I forgot to unload the old zfs module and load the patched version. waiting for import...

**Comment by behlendorf** [comment]
I've added the patch posted here to #4186 so it's easier to find if needed.

**Comment by rlaager** [comment]
I'm duplicating my comment from #4186 for the record:

The consensus seems to be for dropping this in favor of Pavel Zakharov's pending work, as discussed in the "SPA import and pool recovery" lightning talk.

**Comment by behlendorf** [comment]
This issue will be addressed by https://github.com/openzfs/openzfs/pull/539.

**Comment by behlendorf** [comment]
Resolved by 6cb8e5306d9696d155ae7a808f56c4e46d69b64c
